### PR TITLE
GH-776: DataLoader for in-memory datasets

### LINKED ISSUE #776: Performance issue with "make_label_dictionary"
**Describe the bug**
I have a corpus of following split, for a binary classification problem.
Corpus: 1082924 train + 360975 dev + 360975 test sentences
When I run "corpus.make_label_dictionary()", it takes for ever to complete

**Expected behavior**
Labels should be extracted rather quickly

**Screenshots**
Attached below is the profiler output on where the code is taking too long
Timer unit: 1e-06 s

Total time: 489.987 s
File: /opt/conda/lib/python3.6/site-packages/flair/data.py
Function: make_label_dictionary at line 809

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   809                                               def make_label_dictionary(self) -> Dictionary:
   810                                                   """
   811                                                   Creates a dictionary of all labels assigned to the sentences in the corpus.
   812                                                   :return: dictionary of labels
   813                                                   """
   814         1  489987462.0 489987462.0    100.0          labels = set([label.value for sent in self.train for label in sent.labels])
   815                                                   log.info(labels)
   816                                                   label_dictionary: Dictionary = Dictionary(add_unk=False)
   817                                                   for label in labels:
   818                                                       label_dictionary.add_item(label)
   819                                           
   820                                                   max_labels = max([len(sent.labels) for sent in self.train])
   821                                                   if max_labels > 1:
   822                                                       label_dictionary.multi_label = True
   823                                           
   824                                                   return label_dictionary

Version [e.g. flair-0.4.2]:

### PR DESCRIPTION
**Pull request by alanakbik**
This PR introduces a new `DataLoader` class that inherits from PyTorch `DataLoader` and contains optimizations for Flair. Specifically, in cases in which a `Dataset` is in memory, it returns direct pointers to `Sentence` objects of a dataset, allowing us to retain embeddings between epochs in training. This may greatly speed up training for datasets that fit in memory. 

It also modifies the `make_label_dictionary()` method to make use of the DataLoader, to speed up computation (i.e. closes #776), and adds a tqdm progress bar.

### CODE REVIEW

**Comment by kashif** [comment]
:+1:

**Comment by alanakbik** [comment]
:+1:
